feature/session/src/main/java/io/github/droidkaigi/confsched2019/session/ui/widget/TimetableLayoutManager.kt

```diff
@@ -64,35 +93,57 @@ class TimeTableLayoutManager(
     private var firstStartUnixMin = NO_TIME
     private var lastEndUnixMin = NO_TIME
 
+    private var saveState: SaveState? = null
+
     override fun generateDefaultLayoutParams(): RecyclerView.LayoutParams {
         return RecyclerView.LayoutParams(
             RecyclerView.LayoutParams.WRAP_CONTENT,
             RecyclerView.LayoutParams.WRAP_CONTENT
         )
     }
 
+    override fun onRestoreInstanceState(state: Parcelable?) {
+        saveState = (state as? SaveState)
+    }
+
+    override fun onSaveInstanceState(): Parcelable? {
+        if (childCount == 0) return null
+
+        val view = findFirstVisibleView() ?: return null
+        return SaveState(
+            view.adapterPosition,
+            getDecoratedLeft(view),
+            getDecoratedTop(view)
+        )
+    }
+
     override fun onLayoutChildren(recycler: Recycler, state: State) {
         if (itemCount == 0) {
             detachAndScrapAttachedViews(recycler)
             periods.clear()
             columns.clear()
             anchor.reset()
+            saveState = null
             return
         }
 
-        detachAndScrapAttachedViews(recycler)
         anchor.reset()
         calculateColumns()
 
-        var offsetX = parentLeft
-        for (columnNum in 0 until columns.size()) {
-            if (columns.indexOfKey(columnNum) == 0) anchor.leftColumn = columnNum
-            offsetX += fillColumnHorizontally(columnNum, 0, offsetX, parentTop, true, recycler)
-            if (offsetX > parentRight) {
-                anchor.rightColumn = columnNum
-                break
-            }
-        }
+        val firstVisibleView = if (childCount > 0) findFirstVisibleView() else null
+        val offsetX = saveState?.left ?: firstVisibleView?.let(this::getDecoratedLeft)
+        val offsetY = saveState?.top ?: firstVisibleView?.let(this::getDecoratedTop)
+        val period = saveState?.position?.let(periods::getOrNull)
+            ?: firstVisibleView?.adapterPosition?.let(periods::getOrNull)
+        detachAndScrapAttachedViews(recycler)
+        if (offsetX != null && offsetY != null && period != null)
+            relayoutChildren(offsetX, offsetY, period, recycler)
```

Comment:
Don't we need to round offsets like `Math.max(offsetX, parentLeft)` and/or `Math.max(offsetY, parentTop)`?

Reply:
in this case, the first child view has negative `left`(and `top`)

<img src="https://user-images.githubusercontent.com/9051623/51975502-bf690980-24c5-11e9-872c-8a3c3c6e9821.png" width="300px" />